### PR TITLE
Properties Modernization

### PR DESCRIPTION
**Pull request by GrigoriiSolnyshkin**
This pull request
1. removes receiver dependent name mangling for non-private properties;
2. fixes bugs related to open fields and complicated cases of overriding (`val` -> `var`, different types);
3. "immutable" predicates are inhaled rather than being required.

The field is used for property iff it is final and has `FirDefaultPropertyGetter`, `FirDefaultPropertySetter` or both.
This pull request does not try to use fields when type of receiver is final (and property itself is not) although it is a possible further step.

You might want to take a look at new tests: `good_contracts/multiple_interfaces`, `no_contracts/multiple_interfaces`, `good_contracts/private_properties`.

```kotlin
interface Base
fun takeBase(base: Base) { }

class Impl : Base
fun createImpl() {
  val impl = Impl()
  takeBase(impl)
}
```
Note that 3. is not a final solution, but rather a workaround. It is needed because otherwise the example above won't have correct permissions. In case we want to add mutable fields with `write` accesses to this predicate, we won't be able to inhale permissions to them twice. Places where changes have been made are marked with `//TODO (inhale vs require)`.

### CODE REVIEW

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt [hunk 1]
```diff
@@ -40,7 +42,7 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
     ProgramConversionContext {
     private val methods: MutableMap<MangledName, FunctionEmbedding> = SpecialKotlinFunctions.byName.toMutableMap()
     private val classes: MutableMap<MangledName, ClassTypeEmbedding> = mutableMapOf()
-    private val properties: MutableMap<MangledName, PropertyEmbedding> = mutableMapOf()
+    private val properties: MutableMap<CallableId, PropertyEmbedding> = mutableMapOf()
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:45]: Why the change here?  Our mangled names should already be able to uniquely represent callable ids, but I could imagine we may want to have names that are synthetic, too.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt [hunk 2]
```diff
@@ -116,7 +124,7 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
 
                 // Phase 3
                 val properties = symbol.propertySymbols
-                newEmbedding.initFields(properties.mapNotNull { processBackingFields(it, newEmbedding) }.toMap())
+                newEmbedding.initFields(properties.mapNotNull { processBackingFields(it, symbol) }.toMap())
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:127]: Was this a bug, or is this a behaviour change?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt [hunk 3]
```diff
@@ -17,69 +16,192 @@ fun VariableEmbedding.sameSize(): ExpEmbedding =
...
-                    ),
-                )
+sealed interface StdLibReceiverInterface {
+    fun match(function: NamedFunctionSignature): Boolean
+}
+
+sealed interface PresentInterface : StdLibReceiverInterface {
+    val interfaceName: String
+    override fun match(function: NamedFunctionSignature): Boolean =
+        function.receiverType?.isInheritorOfCollectionTypeNamed(interfaceName) == true
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:26]: nit: `b == true` is the same as just `b`

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt [hunk 4]
```diff
@@ -17,69 +16,192 @@ fun VariableEmbedding.sameSize(): ExpEmbedding =
...
+
+data object ListInterface : PresentInterface {
+    override val interfaceName = "List"
+}
+
+data object MutableListInterface : PresentInterface {
+    override val interfaceName = "MutableList"
+}
+
+data object NoInterface : StdLibReceiverInterface {
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:41]: I'm a bit confused, what is the goal of this?

**Review by jesyspa — COMMENTED** [review]
Some brief notes, will take a closer look later

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt (hunk 3 above)
Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:26]: Well, the left part of comparison here has `Boolean?` type. But may be `?: false` is clearer indeed.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt (hunk 2 above)
Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:127]: I just needed a class symbol to check its finality in the function, but I don't need the `TypeEmbedding` now.
I wasn't sure if `FirPropertySymbol.isFinal` always returns true if class is final.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt (hunk 4 above)
Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:41]: It was created for the `emptyList` function which is not a member function.

Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:41]: But I think in the end we have too many problems with this __always writable__ field.
For example, we can't normally access `Map.entries` twice because we will have permission to `Set.size` field of the same `Set` twice and so on.

When we have some uniqueness system implemented, for we will probably be able reason about immutable lists since we have quite a lot of methods to create "correct" and unique (or at least without refs to mutable versions) lists (including `listOf`, `emptyList`, `toList`, `filter` and so on).

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt (hunk 3 above)
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:26]: Ah, I see.  Yeah, I tihnk `?: false` would be better

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt [hunk 5]
```diff
@@ -250,16 +258,28 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
...
+    ): Pair<SimpleKotlinName, FieldEmbedding>? {
+        val embedding = embedClass(classSymbol)
         val unscopedName = symbol.callableId.embedUnscopedPropertyName()
-        val name = symbol.callableId.embedMemberPropertyName()
-        embedding.findAncestorField(unscopedName)?.let { return null }
-        val backingField = name.specialEmbedding() ?: symbol.hasBackingField.ifTrue {
+        val scopedName = symbol.embedMemberPropertyName()
+        val fieldIsAllowed = symbol.hasBackingField
+                && (symbol.getterSymbol?.fir is FirDefaultPropertyGetter ||
+                symbol.setterSymbol?.fir is FirDefaultPropertySetter)
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:278]: I think the logic here needs to be more complicated.  Let's say the getter is default but the setter is custom, then we should not use a field.  But if the getter is default and the setter is missing, we should.  Or is this covered somehow?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt [hunk 6]
```diff
@@ -250,16 +258,28 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
...
+    private fun FirPropertySymbol.embedMemberPropertyName() = this.callableId.embedMemberPropertyName(
+        Visibilities.isPrivate(this.visibility)
+    )
+
     /**
      * Construct and register the field embedding for this property's backing field, if any exists.
      */
-    private fun processBackingFields(symbol: FirPropertySymbol, embedding: ClassTypeEmbedding): Pair<SimpleKotlinName, FieldEmbedding>? {
+    @OptIn(SymbolInternals::class)
+    private fun processBackingFields(
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:269]: nit: I think the function should have been renamed to `processBackingField` (singular) at some point.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt (hunk 4 above)
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:41]: Ah, I see.  If you need to, feel free to disable support for it for now.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt [hunk 7]
```diff
@@ -17,69 +16,192 @@ fun VariableEmbedding.sameSize(): ExpEmbedding =
...
-            ifFunctionName("add") {
-                return listOf(receiver!!.increasedSize(1))
-            }
-            ifFunctionName("isEmpty") {
-                return listOf(
-                    receiver!!.sameSize(),
-                    Implies(returnVariable, EqCmp(FieldAccess(receiver, ListSizeFieldEmbedding), IntLit(0))),
-                    Implies(Not(returnVariable), GtCmp(FieldAccess(receiver, ListSizeFieldEmbedding), IntLit(0)))
-                )
+sealed interface StdLibPreCondition : StdLibCondition {
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt:67]: nit: `Precondition` (it's one word), same for `Postcondition` elsewhere

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/names/NameEmbeddings.kt [hunk 8]
```diff
@@ -37,7 +37,16 @@ fun CallableId.embedExtensionSetterName(type: TypeEmbedding): ScopedKotlinName =
     embedScopedWithType(type, ExtensionSetterKotlinName(callableName))
 
 fun CallableId.embedLocalPropertyName(): KotlinName = SimpleKotlinName(callableName)
-fun CallableId.embedMemberPropertyName(): ScopedKotlinName = embedScoped(MemberKotlinName(callableName))
+fun CallableId.embedMemberPropertyName(isPrivate: Boolean): ScopedKotlinName {
+    val id = classId ?: error("Embedding non-member property $callableName as a member.")
+    val className = ClassKotlinName(id.relativeClassName)
+    val packageName = packageName
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/names/NameEmbeddings.kt:43]: nit: unnecessary?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionContext.kt [hunk 9]
```diff
@@ -71,19 +76,60 @@ fun StmtConversionContext.declareLocalVariable(symbol: FirVariableSymbol<*>, ini
...
+ */
+fun FirPropertySymbol.findFinalParentProperty(): FirPropertySymbol? =
+    if (this !is FirIntersectionOverridePropertySymbol)
+        isFinal.ifTrue { this }
+    else propertyIntersections.firstNotNullOfOrNull { it.findFinalParentProperty() }
+
+fun FirPropertySymbol.findSomeParentProperty(): FirPropertySymbol =
+    if (this is FirIntersectionOverridePropertySymbol)
+        propertyIntersections.first().findSomeParentProperty()
+    else this
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionContext.kt:100]: I'd like to understand what's going on here a bit better; why do we need to find the final property rather than working with the one that's here?

**Review by jesyspa — COMMENTED** [review]
I think I'd like to get a higher-level understanding of our approach to this.  In particular:
1. I understand that distinguishing between fields of different classes isn't useful anymore, so we can bunch all of those together.
2. But when we have a property, we want to look it up to resolve it to a field or a getter/setter.  But it's not quite clear to me how this is achieved now.
3. I do believe it works since it seems essential for the `multiple_interfaces` test.

A good solution might be to document how we represent these things as e.g. comments in `ProgramConverter` or `ClassEmbedding`.

**Comment by francescoo22** [comment]
Since inhaling predicates instead of requiring them is just a temporary solution, it would be useful to point it out, document  why the decision has been made and which problems we have while requiring predicates (perhaps you can create an issue on space).
Moreover, since the PR modifies a lot of files, it would be useful to write in the description which parts have been modified to inhale predicates, so that it will be easier for me to revert it in the future.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt (hunk 5 above)
Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:278]: I guess it is not that important how to deal with the cases when the getter is default and the setter is custom or vice versa. But I agree it can create unwanted behavior with current implementation (e.g. inhale access and then call the getter method).

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/names/NameEmbeddings.kt (hunk 8 above)
Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/names/NameEmbeddings.kt:43]: It is still necessary for the private fields and regarding public fields it is necessary to leave package and class in order to obtain correct comparison behavior.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionContext.kt (hunk 9 above)
Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionContext.kt:100]: Overall this is the part I'd like to discuss with you and Ilya during this week.

Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionContext.kt:100]: Thanks, the comments help.  Let's discuss with Ilya.  My problem here is: we are doing some work to go from a specific piece of information we have, to some arbitrary choice.

One thing to note here: when a function takes an interface parameter, that can't possible contain any access permissions.  So for an interface property, we don't actually care which interface we take it from.  But perhaps we can do something better than taking it from a random one?  One idea is that we could have a common method for the actual call, for example, and then use the property symbol we have here to find the right access to inhale?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/names/NameScope.kt [hunk 10]
```diff
@@ -27,11 +27,25 @@ data class GlobalScope(override val packageName: FqName) : PackagePrefixScope {
...
+data class DefaultClassScope(override val packageName: FqName, override val className: ClassKotlinName, ) : ClassScope {
     override val suffix = className.mangled
+}
+
+/**
+ * We do not want to mangle field names with class and package, hence introducing
+ * this special `NameScope`. Note that it still needs package and class for other purposes.
+ */
+data class PropertyClassScope(override val packageName: FqName, override val className: ClassKotlinName) : ClassScope {
+    override val suffix = className.mangled + "_property"
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/names/NameScope.kt:43]: nit: I think `backing_field` would be better when that's what we generate, with `property` used for getter/setter

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/ControlFlow.kt [hunk 11]
```diff
@@ -207,7 +207,8 @@ data class FunctionExp(val signature: FullNamedFunctionSignature?, val body: Exp
             // Ideally we would want to assume these rather than inhale them to prevent inconsistencies with permissions.
             // Unfortunately Silicon for some reason does not allow Assumes. However, it doesn't matter as long as the
             // provenInvariants don't contain permissions.
-            arg.provenInvariants().forEach { invariant ->
+            // TODO (inhale vs require) Decide if `predicateAccessInvariant` should be required rather than inhaled in the beginning of the body.
+            (arg.provenInvariants() + listOfNotNull(arg.predicateAccessInvariant())).forEach { invariant ->
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/ControlFlow.kt:211]: Is this change necessary for this PR to work?  If not, I would recommend not including it.

Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/ControlFlow.kt:211]: Without this change I can't add any tests where function call accepts some subtype of the type for which it was defined, only the type itself. I thought it was important to show that implementation of the properties works in that case.

Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/ControlFlow.kt:211]: Is that because more work is needed on unfolding in that case?

Comment (GrigoriiSolnyshkin) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/ControlFlow.kt:211]: You kind of need to unfold the predicate until the necessary type is met.

**Review by jesyspa — APPROVED** [review]
LGTM
